src/nomad_simulations/properties/band_gap.py

```diff
@@ -0,0 +1,147 @@
+#
+# Copyright The NOMAD Authors.
+#
+# This file is part of NOMAD. See https://nomad-lab.eu for further info.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import numpy as np
+from structlog.stdlib import BoundLogger
+import pint
+from typing import Optional
+
+from nomad.units import ureg
+from nomad.metainfo import Quantity, MEnum, Section, Context
+
+from ..physical_property import PhysicalProperty
+
+
+class ElectronicBandGap(PhysicalProperty):
+    """
+    Energy difference between the highest occupied electronic state and the lowest unoccupied electronic state.
+    """
+
+    iri = 'https://fairmat-nfdi.github.io/fairmat-taxonomy/#/ElectronicBandGap'
+
+    # ? can `type` change character depending on the `variables`?
+    type = Quantity(
+        type=MEnum('direct', 'indirect'),
+        description="""
+        Type categorization of the `ElectronicBandGap`. This quantity is directly related with `momentum_transfer` as by
+        definition, the electronic band gap is `'direct'` for zero momentum transfer (or if `momentum_transfer` is `None`) and `'indirect'`
+        for finite momentum transfer.
+
+        Note: in the case of finite `variables`, this quantity refers to all of the `value` in the array.
+        """,
+    )
+
+    momentum_transfer = Quantity(
+        type=np.float64,
+        shape=[2, 3],
+        description="""
+        If the `ElectronicBandGap` is `'indirect'`, the reciprocal momentum transfer for which the band gap is defined
+        in units of the `reciprocal_lattice_vectors`. The initial and final momentum 3D vectors are given in the first
+        and second element. Example, the momentum transfer in bulk Si2 happens between the Γ and the (approximately)
+        X points in the Brillouin zone; thus:
+            `momentum_transfer = [[0, 0, 0], [0.5, 0.5, 0]]`.
+
+        Note: this quantity only refers to scalar `value`, not to arrays of `value`.
+        """,
+    )
+
+    spin_channel = Quantity(
+        type=np.int32,
+        description="""
+        Spin channel of the corresponding `ElectronicBandGap`. It can take values of 0 or 1.
+        """,
+    )
```

Comment:
In this structure, both channels live separately, as elements of a list (`band_gap[0]` and `band_gap[1]`). The character of the state is determined by a complete different property (actually, by the **sign** of another physical property, the `magnetization`).

I prefer to keep 2 elements for `band_gap`. My argument also goes in the direction of keeping `variables` defined for each channel independently, to avoid potential issues. Your suggestion also implies having to restructure the shape of a `value` in a more complex way.

Reply:
Well, my last suggestion is an alternative to the one above. I like that one better, tbh, and it's also fully compatible with your suggestion here.

Still, **you didn't really address my questions / concerns**:

- where are you communicating _spin-restriction_? Does a user always have to analyze the methodologies first before tackling the output?

1. While this might generally be the proper way of working, I'm not convinced everyone will comply. I think a slight redundancy is fine, if it eases the data handling.
2. Our queries (especially GrapQL) enables _targeted data extraction_. If spin-restriction isn't communicated in some way or shape in the output, users will also have to download (part of) the method.

- how do you communicate that 2 spin-channels have the same band gap / DOS, i.e. the material exhibits _no spontaneous spin-polarization_? Again, I think we should a_void that users have to build out their own analysis for something so basic_.

_That last phrase is really the determining factor for me._ Overall, I agree with your approach, though, and want to stay clear from nesting. Still, one has to be a pragmatist, not an ideolog.

Reply:
> where are you communicating spin-restriction? Does a user always have to analyze the methodologies first before tackling the output?

I'd argue that if it is a special case for only treating one spin_channel and not the other, indeed, there should be a flag indicating this, and spin_channel could be set to 0. 
But that does not mean the user needs to "analyze the methodologies first before tackling the output". I mean, for what? This is only an indicator of a calculation done in certain spin channels, hence `spin_channel` should already tell you about it.

> how do you communicate that 2 spin-channels have the same band gap / DOS, i.e. the material exhibits no spontaneous spin-polarization?

What? There is truly only 2 situations:
1. Simulation is not spin-polarized: then there is only one element for band gap and spin_channel is not assigned.
2. Simulation is spin polarized: there are 2 elements band gap and spin_channel is assign.

Now, if band gap is equal in both spin channels, it means that the state is either AFM or PM. Nobody is going to avoid having to go to another property (`magnetization`) and check the nature of the state.

___ 

In any case, I am not sure what you meant. I am not 100% sure your example above is realistic either.
Btw, this has nothing to do with me not understanding you, but rather on how you communicate this (so there is no need of "putting this in any other way"). Just a couple of suggestions (don't feel I am trying to be harsh, it is just that it gets very complex and lengthy any kind of review, and I'd like to keep it sharp to the point):
- Please, stay away from abusing the use of **bold** and _italics_. It truly makes your messages unreadable.  Furthermore, I (and probably others as well) understand what you mean in each phrase, as my brain is capable of emphasizing without the need of bolding or doing whole phrases in italics, so you should not worry about that but rather:
- Convey in defining a clear and concise idea in one or two phrases before[1] doing the review. If your messages are this long, and they keep being long reply after reply, it means that something is not going on properly. My suggestion would be that you re-summarize the message before sending it, not once, but a couple of times (this is also a problem I have btw, maybe I am overdoing it to the point my messages are cryptic, but I am sure there should be an intermediate point).


[1] See, I didn't need to bold **before** to put the emphasis, as I know you are more than capable of understanding where the emphasis lies in that phrase.

Reply:
> Now, if band gap is equal in both spin channels, it means that the state is either AFM or PM. Nobody is going to avoid having to go to another property (magnetization) and check the nature of the state.

I'm fine if you want to convey that information in another property, but then I would add it to this commit: a PM state is easily checked for.

The reason I'm mentioning this is that users will want to easily search and understand whether a property was spin-polarized because it was artificially constrained, or as an actual outcome of the setup.

Reply:
A PM state is indistinguishable from AFM only by looking at BandGap. Maybe we can leave it for when we have a proper example and the `magnetization` defined?